drivers/amlogic/media/vout/hdmitx/hdmi_tx_20/hdmi_tx_main.c

```diff
@@ -6162,10 +6131,14 @@ static void hdmitx_hpd_plugout_handler(struct work_struct *work)
 		mutex_unlock(&setclk_mutex);
 		return;
 	}
-	/*after plugout, DV mode can't be supported*/
-	hdmitx_set_drm_pkt(NULL);
-	hdmitx_set_vsif_pkt(0, 0, NULL, true);
-	hdmitx_set_hdr10plus_pkt(0, NULL);
+	if (is_tv_changed()) {
```

Comment:
I don't think that this is correct.
TV may change on plugin, not plugout.
I think that this will always return false, and never be called.

Reply:
It's called again on the plugin.  
It appears TV change triggers on the plugout as well.  I kept this in my test build for extra debugging:
	if (is_tv_changed()) {
		/* after plugout, DV mode can't be supported
		 * clear pkt
		 */
		pr_info("hdmitx_hpd_plugout_handler TV changed - line 6138\n");
		hdmitx_set_drm_pkt(NULL);
		hdmitx_set_vsif_pkt(0, 0, NULL, true);
		hdmitx_set_hdr10plus_pkt(0, NULL);
	}

When I physically remove the HDMI from my box while playing I get this:

```
[  801.569798@0]h hdmitx: system: irq 80000004 0
[  801.569840@0]- hdmitx: system: update rx hdr info d
[  801.569842@0]- hdmitx: system: plugout
[  801.569845@0]- hdmitx: hw: hdmitx_get_isaformat:0x0
[  801.569847@0]- hdmitx: hw: P_HHI_HDMI_CLK_CNTL :0x10100
[  801.569849@0]- hdmitx: hw: P_HHI_HDMI_PLL_CNTL :0xdb0004b9
[  801.569861@0]- hdmitx: system: update physical size: 16 9
[  801.581641@0]h hdmitx: hdmitx_set_drm_pkt: tf=1, cf=1, colormetry=0
[  802.093279@0]h hdmitx: system: irq 80000002 0
[  802.622773@0]- hdmitx: system: plugin
[  802.650557@0]- hdmitx: hw: ddc timeout
[  802.670898@0]- hdmitx: hw: ddc timeout
[  802.695236@0]- hdmitx: hw: ddc timeout
[  802.715483@0]- hdmitx: hw: ddc timeout
[  802.735849@0]- hdmitx: hw: ddc timeout
[  802.756213@0]- hdmitx: hw: ddc timeout
[  802.837092@0]h hdmitx: system: irq 80000004 0
[  802.919698@0]d hdmitx: edid: EDID Parser:
[  802.919703@0]d hdmitx: edid: PLUGIN_DVI_OUT
[  802.919706@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919711@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919713@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919715@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919716@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919718@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919719@0]d hdmitx: edid: unkonw Extension Tag detected, continue
[  802.919742@0]d hdmitx: edid: not find IEEEOUT
[  802.919744@0]d hdmitx: edid: check sum invalid
[  802.919758@0]- hdmitx: edid: 62 errors between two reading
[  802.919760@0]- hdmitx: edid: check sum invalid
[  802.919765@0]- hdmitx: edid: blk0 raw data
[  802.919790@0]- hdmitx: edid: 
[  802.919790@0]- 0fffffffffffffff15af1670ffffffff22af1670ffffffff0aee91a3544c9926
[  802.919790@0]- 0f5054a1544c9926455054a1544c9926d1c054a1544c9926d1c054a1544c9926
[  802.919790@0]- 8a0fffffffffffff6fc203ffffffffff307fffffffffffff03ffffffffffffff
[  802.919790@0]- 781eff3c00ffffff20ffffffffffffff07ffffffffffffff20ffffffffffffff
[  802.919790@0]- 
[  802.919790@0]- 
[  802.919792@0]- hdmitx: edid: check sum invalid
[  802.919795@0]- hdmitx: edid: blk0 raw data
[  802.919818@0]- hdmitx: edid: 
[  802.919818@0]- 20ffffffffffffff20ffffffffffffff20ffffffffffffff20ffffffffffffff
[  802.919818@0]- 20ffffffffffffff20ffffffffffffff20ffffffffffffff20ffffffffffffff
[  802.919818@0]- 20ffffffffffffff20ffffffffffffff20ffffffffffffff20ffffffffffffff
[  802.919818@0]- 20ffffffffffffff20ffffffffffffff20ffffffffffffff20ffffffffffffff
[  802.919818@0]- 
[  802.919818@0]- 
[  802.919821@0]- hdmitx: system: update physical size: 2550 2550
[  802.919825@0]- hdmitx: system: get current mode: 2160p24hz
[  802.919827@0]- hdmitx: system: update physical size: 2550 2550
[  802.919842@0]- hdmitx: hdmitx: display colour subsampling is auto set to YUV444 (VIC: 93)
[  802.919844@0]- hdmitx: hdmitx: display colourdepth is auto set to 8 bits (VIC: 93)
[  802.919868@0]- hdmitx: video: already init VIC = 0  Now VIC = 93
[  802.919870@0]- hdmitx: video: rx edid only support RGB format
[  802.919873@0]- hdmitx: hw: div40: 0
[  802.948772@0]- hdmitx: E: ddc w1b 0x54 0x20 0x00
[  802.948780@0]- hdmitx: frac_rate = 0
[  802.948786@0]- hdmitx: config HPLL = 5940000 frac_rate = 0
[  802.948840@0]- hdmitx: HPLLv1: 0xdb3a04f7
[  802.948842@0]- hdmitx: config HPLL done
[  802.948847@0]- hdmitx: set_hpll_od3_clk_div[687] div = 6
[  802.948852@0]- hdmitx: j = 8  vid_clk_div = 2
[  802.948864@0]- hdmitx: set enc for VIC: 93
[  802.990780@0]- hdmitx: hw:  config hdmitx IP vic = 93 cd:4 cs: 0
[  803.000877@0]- hdmitx: video: Sink is HDMI device
[  803.000905@0]- hdmitx: hw: set audio
[  803.000911@0]- hdmitx: hw: hdmitx tx_aud_src = 0, audio_param->channel_num = 1, hdev->aud_output_ch = 0
[  803.000923@0]- hdmitx: fs = 7, cd = 4, tmds_clk = 297000
[  803.000925@0]- hdmitx: hw: aud_n_para = 20480
[  803.000933@0]- hdmitx: hw: set channel status
[  803.000945@0]- hdmitx: hw: [0x10e3] = 0x1f
[  803.000964@0]- hdmitx: audio: Audio Type: Dobly Digital +
[  803.000967@0]- hdmitx: system: update rx hdr info 0
[  803.000968@0]- hdmitx: hw: set audio
[  803.000970@0]- hdmitx: hw: hdmitx tx_aud_src = 0, audio_param->channel_num = 1, hdev->aud_output_ch = 0
[  803.000981@0]- hdmitx: fs = 7, cd = 4, tmds_clk = 297000
[  803.000982@0]- hdmitx: hw: aud_n_para = 20480
[  803.000989@0]- hdmitx: hw: set channel status
[  803.001001@0]- hdmitx: hw: [0x10e3] = 0x1f
[  803.001019@0]- hdmitx: audio: Audio Type: Dobly Digital +
[  803.001149@0]- hdmitx: system: update rx hdr info 0
[  803.001166@0]h hdmitx: system: irq 80000001 0
[  803.001173@0]- hdmitx: system: plugout
[  803.001176@0]- hdmitx: hw: hdmitx_get_isaformat:0x0
[  803.001178@0]- hdmitx: hw: P_HHI_HDMI_CLK_CNTL :0x100
[  803.001180@0]- hdmitx: hw: P_HHI_HDMI_PLL_CNTL :0xdb0104f7
**[  803.001182@0]- hdmitx: hdmi crc is diff between uboot and kernel
[  803.001183@0]- hdmitx: hdmitx_hpd_plugout_handler TV changed - line 6138
[  803.001186@0]d hdmitx: hdmitx_set_drm_pkt: disable drm pkt
[  803.001195@0]d hdmitx: TV not support DV, clr dv_vsif**
[  803.001204@0]d hdmitx: hdmitx: H14b VSIF, switching signal to SDR
[  803.001219@0]- hdmitx: hdmitx_set_hdr10plus_pkt: null vsif
[  803.001261@0]- hdmitx: system: update physical size: 16 9
[  803.102741@0]- hdmitx: [0x10000103]: 0x00000008
[  803.102748@0]- hdmitx: [0x10000170]: 0x00000010
[  803.102750@0]- hdmitx: [0x10000180]: 0x000000ff
[  803.102752@0]- hdmitx: [0x10000181]: 0x000000ff
[  803.102754@0]- hdmitx: [0x10000182]: 0x00000003
[  803.102756@0]- hdmitx: [0x10000183]: 0x00000007
[  803.102758@0]- hdmitx: [0x10000184]: 0x0000003f
[  803.102760@0]- hdmitx: [0x10000185]: 0x00000002
[  803.102762@0]- hdmitx: [0x10000187]: 0x000000ff
[  803.102764@0]- hdmitx: [0x10000188]: 0x00000003
[  804.895590@0]h hdmitx: system: irq 80000002 0
[  805.406750@0]- hdmitx: system: plugin
[  805.521284@0]d hdmitx: edid: EDID Parser:
[  805.521305@0]d hdmitx: Edid_ParsingVendSpec:ieeeoui=0xd046,len=11
[  805.521309@0]d hdmitx: v2 VSVDB: len=11, sup_2160p60hz=1, Interface=2
[  805.521310@0]d hdmitx: Edid_ParsingVendSpec:ieeeoui=0xd046,len=11
[  805.521312@0]d hdmitx: v2 VSVDB: len=11, sup_2160p60hz=1, Interface=2
[  805.521330@0]d hdmitx: edid: get dtd0 vic: 97
[  805.521334@0]d hdmitx: hdmitx: get PMT vic: 97
[  805.521337@0]d hdmitx: edid: find IEEEOUT
[  805.521351@0]- hdmitx: edid: blk0 raw data
[  805.521376@0]- hdmitx: edid: 
[  805.521376@0]- 00ffffffffffff0015af1670000e00012221010380a05a780aee91a3544c9926
[  805.521376@0]- 0f5054a1080031404540614071408180d1c00101010108e80030f2705a80b058
[  805.521376@0]- 8a0040846300001e6fc200a0a0a055503020350040846300001e000000fd0018
[  805.521376@0]- 781eff3c000a202020202020000000fc004d52310a2020202020202020200155
[  805.521376@0]- 
[  805.521376@0]- 
[  805.521378@0]- hdmitx: edid: blk1 raw data
[  805.521401@0]- hdmitx: edid: 
[  805.521401@0]- 02036df05a61606665101f0413051403021220212215015d5e5f6263643f4035
[  805.521401@0]- 0f7f07677e035706031507503d1fc05f7e035f7e01834f00006e030c002600b8
[  805.521401@0]- 3c240080010203046ad85dc401788003022878e200cfe305c000e3060d01e20f
[  805.521401@0]- ffeb0146d0004803768a66727c000000000000000000000000000000000000b6
[  805.521401@0]- 
[  805.521401@0]- 
[  805.521450@0]- hdmitx: system: update physical size: 1600 900
[  805.521454@0]- hdmitx: system: get current mode: 2160p24hz
[  805.521456@0]- hdmitx: system: update physical size: 1600 900
[  805.521470@0]- hdmitx: hdmitx: display colour subsampling is auto set to YUV444 (VIC: 93)
[  805.521472@0]- hdmitx: hdmitx: display colourdepth is auto set to 12 bits (VIC: 93)
[  805.521493@0]- hdmitx: video: already init VIC = 0  Now VIC = 93
[  805.521496@0]- hdmitx: hw: div40: 1
[  805.543604@0]h hdmitx: hdmitx_set_drm_pkt: HDR->SDR, DRM_DB[0]=0
[  805.543910@0]- hdmitx: hdr_work_func: send zero DRM
[  805.549059@0]- hdmitx: frac_rate = 0
[  805.549068@0]- hdmitx: config HPLL = 4455000 frac_rate = 0
[  805.549123@0]- hdmitx: HPLL: 0xdb0004b9
[  805.549124@0]- hdmitx: config HPLL done
[  805.549130@0]- hdmitx: set_hpll_od3_clk_div[687] div = 10
[  805.549135@0]- hdmitx: j = 9  vid_clk_div = 1
[  805.549147@0]- hdmitx: set enc for VIC: 93
[  805.590776@0]- hdmitx: hw:  config hdmitx IP vic = 93 cd:6 cs: 2
[  805.590876@0]h hdmitx: system: irq 80000001 0
[  805.601823@0]- hdmitx: video: Sink is HDMI device
[  805.601849@0]- hdmitx: hw: set audio
[  805.601858@0]- hdmitx: hw: hdmitx tx_aud_src = 0, audio_param->channel_num = 1, hdev->aud_output_ch = 0
[  805.601871@0]- hdmitx: fs = 7, cd = 6, tmds_clk = 297000
[  805.601873@0]- hdmitx: hw: aud_n_para = 24576
[  805.601881@0]- hdmitx: hw: set channel status
[  805.601893@0]- hdmitx: hw: [0x10e3] = 0x1f
[  805.601912@0]- hdmitx: audio: Audio Type: Dobly Digital +
[  805.601915@0]- hdmitx: system: update rx hdr info d
[  805.601917@0]- hdmitx: hw: set audio
[  805.601918@0]- hdmitx: hw: hdmitx tx_aud_src = 0, audio_param->channel_num = 1, hdev->aud_output_ch = 0
[  805.601929@0]- hdmitx: fs = 7, cd = 6, tmds_clk = 297000
[  805.601930@0]- hdmitx: hw: aud_n_para = 24576
[  805.601937@0]- hdmitx: hw: set channel status
[  805.601949@0]- hdmitx: hw: [0x10e3] = 0x1f
[  805.601967@0]- hdmitx: audio: Audio Type: Dobly Digital +
[  805.602851@0]h hdmitx: hdmitx_set_drm_pkt: tf=16, cf=9, colormetry=0
[  805.694730@0]- hdmitx: [0x10000103]: 0x00000008
[  805.694735@0]- hdmitx: [0x10000104]: 0x00000001
[  805.694739@0]- hdmitx: [0x10000170]: 0x00000010
[  805.694741@0]- hdmitx: [0x10000180]: 0x000000ff
[  805.694743@0]- hdmitx: [0x10000181]: 0x000000ff
[  805.694745@0]- hdmitx: [0x10000182]: 0x00000003
[  805.694747@0]- hdmitx: [0x10000183]: 0x00000007
[  805.694748@0]- hdmitx: [0x10000184]: 0x0000003f
[  805.694750@0]- hdmitx: [0x10000185]: 0x00000002
[  805.694753@0]- hdmitx: [0x10000187]: 0x000000ff
[  805.694754@0]- hdmitx: [0x10000188]: 0x00000003
```

Reply:
This is what I've come up with earlier today, I think it's a safer option. What do you think?
```diff
diff --git a/drivers/amlogic/media/vout/hdmitx/hdmi_tx_20/hdmi_tx_main.c b/drivers/amlogic/media/vout/hdmitx/hdmi_tx_20/hdmi_tx_main.c
index 35ac77e20087..02603e06c1a0 100644
--- a/drivers/amlogic/media/vout/hdmitx/hdmi_tx_20/hdmi_tx_main.c
+++ b/drivers/amlogic/media/vout/hdmitx/hdmi_tx_20/hdmi_tx_main.c
@@ -173,6 +173,7 @@ static struct hdmitx_uevent hdmi_events[] = {
 	},
 };
 
+static bool plugout_event;
 /* indicate plugout before systemcontrol boot  */
 static bool plugout_mute_flg;
 static char hdmichecksum[11] = {
@@ -1859,6 +1860,9 @@ void update_current_para(struct hdmitx_dev *hdev)
 	hdev->para = hdmi_get_fmt_name(mode, hdev->fmt_attr);
 }
 
+static struct vsif_debug_save vsif_backup_info = {
+	.signal_sdr = 1
+};
 struct vsif_debug_save vsif_debug_info;
 struct vsif_debug_save hsty_vsif_config_data[8];
 unsigned int hsty_vsif_config_loc, hsty_vsif_config_num;
@@ -6015,6 +6019,13 @@ bool is_tv_changed(void)
 	return ret;
 }
 
+static void backup_vsif_config(struct hdmitx_dev *hdev)
+{
+	vsif_backup_info.type = hdev->hdmi_current_eotf_type;
+	vsif_backup_info.tunnel_mode = hdev->hdmi_current_tunnel_mode;
+	vsif_backup_info.signal_sdr = hdev->hdmi_current_signal_sdr;
+}
+
 static void hdmitx_hpd_plugin_handler(struct work_struct *work)
 {
 	char bksv_buf[5];
@@ -6062,6 +6073,17 @@ static void hdmitx_hpd_plugin_handler(struct work_struct *work)
 	}
 
 	set_disp_mode_auto();
+	if (vsif_backup_info.type != EOTF_T_NULL) {
+		if (!is_tv_changed()) {
+			pr_info(SYS "plugin: Same TV, restore eotf type (%d), tunnel mode (%d) and SDR flag (%d)\n",
+				vsif_backup_info.type, vsif_backup_info.tunnel_mode, vsif_backup_info.signal_sdr);
+			hdmitx_set_vsif_pkt(vsif_backup_info.type, vsif_backup_info.tunnel_mode, NULL, vsif_backup_info.signal_sdr);
+			set_disp_mode_auto();
+		} else {
+			pr_info(SYS "plugin: Different TV, don't restore eotf type, tunnel mode and SDR flag\n");
+		}
+	}
+	plugout_event = false;
 	info = hdmitx_get_current_vinfo();
 	if (info && (info->mode == VMODE_HDMI))
 		hdmitx_set_audio(hdev, &(hdev->cur_audio_param));
@@ -6162,6 +6184,9 @@ static void hdmitx_hpd_plugout_handler(struct work_struct *work)
 		mutex_unlock(&setclk_mutex);
 		return;
 	}
+	if (!plugout_event)
+		backup_vsif_config(hdev);
+	plugout_event = true;
 	/*after plugout, DV mode can't be supported*/
 	hdmitx_set_drm_pkt(NULL);
 	hdmitx_set_vsif_pkt(0, 0, NULL, true);
```

Reply:
Btw with refresh rate switching set to off. My fix attempt behaves strangely on DV titles.  I think you may be onto something here.  It's similar to my attempt, but yes it looks safer.  Will try this out, now.

Reply:
Any findings?